### ISSUE TITLE
CLA signing

### ISSUE THREAD

**Issue by sbellem** [bigchaindb/org#34]
We need some way to manage this so that we can quickly merge new contributors' code.

The ideal would be an automated solution, such as https://cla-assistant.io/ or https://clabot.github.io/

Either way, maintainers need to be able to know whether someone has signed the CLA or not, and when that is not the case, maintainers need to know the procedure to follow in order to on-board a new committer.

**Comment by vrde** [comment]
Having a bot doing this would be great:
 - humans make errors
 - i don't have a full overview on who signed what
 - i don't want to spend time asking around if someone signed the CLA

I think this should be done soon.

**Comment by sbellem** [comment]
> humans make errors

true, and robots don't, e.g.:

![](https://media.giphy.com/media/EizPK3InQbrNK/giphy.gif)

**Comment by sbellem** [comment]
Another example motivating the move to a bot, and clearly showing the absolute greatness and perfection of robots:

![](https://media.giphy.com/media/dKs4NntOpkFRm/giphy.gif)

**Comment by gmcmullen** [comment]
This is all I have to add for now:

![](https://media.giphy.com/media/ZybmZZHlkyPTi/giphy.gif)

Why does NO ONE want to shake her hand?

**Comment by sbellem** [comment]
How about we use BigchainDB to manage the CLA? As a very rough example: once a user digitally signs the CLA, this would trigger a transaction in which the permission to merge code is transferred over to the user.

**Comment by sbellem** [comment]
An example of a somewhat involved automated handling of CLAs: https://github.com/kubernetes/minikube/pull/1295

note that three bots are involved (@googlebot, @k8s-ci-robot, @minikube-bot)

**Comment by kremalicious** [comment]
> How about we use BigchainDB to manage the CLA?

yes, ideal case but: startup, resource & time constraints, yada yada. Solution for now might as well be an excel sheet all maintainers get access to.

For something automated, using https://cla-assistant.io/ looks like a hassle free solution for us and contributors. An external service, is essentially a bot and has a nice unobtrusive flow for contributors, [see here at the end](https://cla-assistant.io/assets/images/preview.gif).

Somebody wants to connect them and create a test PR to see if that works for us? Most importantly, see if we can keep track of signatures within their tool.

@gmcmullen do we require everything we ask for on bigchaindb.com/cla/ to make CLA "signing" legally binding? Cause I assume cla-assistant.io and similar tools capture the GitHub username only.

Also here's another perfectly working robot, gotta adapt in here:
![bender](https://cloud.githubusercontent.com/assets/90316/24369891/c4a40e8a-1325-11e7-83cf-a99cc1e56d11.gif)

**Comment by ttmc** [comment]
As far as I know, CLA-Assistant doesn't support the two options of:

1) agree as an individual,
2) agree on behalf of an entire organization, so that anyone in that organization can make pull requests afterwards.

We've had three organizations do option number 2 so far.

**Comment by ttmc** [comment]
I took a look at CLA-Assistant again just now. It looks like it identifies the contributor by their GitHub account alone, which is _not enough_. A CLA is a legal agreement (contract) and we need more than just a GitHub account (username). Ideally, we get their real name and an email address (but we don't ask them to post those publicly online).

**Comment by ttmc** [comment]
I found out that [CLA-Assistant lets you request additional information from the CLA signer](https://github.com/cla-assistant/cla-assistant#request-more-information-from-the-cla-signer). Good!

As for people who want their entire organization to sign a CLA, I will try to add something to tell people to how use the old process if they want to do that.

**Comment by ttmc** [comment]
As suggested by the README file in [the GitHub repo for CLA-Assistant](https://github.com/cla-assistant/cla-assistant), I created a [new Gist](https://gist.github.com/ttmc/f515d635d6dac1c67130088b4bf4b729) with two files, 1) the Individual CLA and 2) a file named `metadata` containing a JSON Schema file specifying other things to ask:
```json
{
    "name": {
        "title": "Your Full Name",
        "type": "string",
        "githubKey": "name",
        "required": true
    },
    "email": {
        "title": "Your Email",
        "type": "string",
        "githubKey": "email",
        "required": true
    },
    "agreed-and-accepted": {
        "title": "Agreed and accepted: By clicking and accepting this Agreement, I represent and warrant that I have authority to bind the entity named above to the terms and conditions of this Agreement.",
        "type": "boolean",
        "required": true
    },
    "understand-for-individual": {
        "title": "I understand that I am agreeing for myself, not for my organization, and that if I want to agree for my organization, I should go to https://www.bigchaindb.com/cla/ instead.",
        "type": "boolean",
        "required": true
    },
    "agreement": {
        "title": "I have read and agree to the terms of this Agreement.",
        "type": "boolean",
        "required": true
    }
}
```

Then I signed in to CLA Assistant at cla-assistant.io with my @ttmc GitHub account. I connected the bigchaindb/bigchaindb repo with the new Gist.

Then I tested it. It failed.

## Details

I used my @thomasmcco GitHub account to submit a pull request to bigchaindb/bigchaindb from an (apparently) external user:

https://github.com/bigchaindb/bigchaindb/pull/1929

Soon after creating that PR, @CLAassistant posted a comment saying I need to sign the CLA. I clicked the link to do that and did all the steps (as @thomasmcco).

The first time around I got an error message saying something was wrong with the JSON in the `metadata` file, but I couldn't see any problems (and nor could several JSON linters I tried), so I removed the `metadata` file. The second time around everything seemed to work great. I got taken back to the original PR in the end... but it still said:

```text
Some checks haven’t completed yet
1 pending and 1 successful checks
license/cla — Contributor License Agreement is not signed yet.
```

When I clicked the link to "recheck", it took me to a page saying that I _had_ signed the CLA.

CLA Assistant seems to be broken.

**Comment by ttmc** [comment]
Here are some screenshots that I took as I tested CLA Assistant:

![screenshot from 2017-12-04 16-30-34](https://user-images.githubusercontent.com/1210951/33563670-dcc57990-d918-11e7-9266-e5632241dfaf.png)

![screenshot from 2017-12-04 16-31-53](https://user-images.githubusercontent.com/1210951/33563682-e25edfae-d918-11e7-9e39-bfd3c7604e92.png)

![screenshot from 2017-12-04 16-33-08](https://user-images.githubusercontent.com/1210951/33563716-f7ec13b4-d918-11e7-9dfd-548e848762cf.png)

![screenshot from 2017-12-04 16-52-15](https://user-images.githubusercontent.com/1210951/33563726-fc330bd0-d918-11e7-836b-a5b0be8d5cc2.png)

![screenshot from 2017-12-04 16-52-51](https://user-images.githubusercontent.com/1210951/33563735-014b9db2-d919-11e7-9f37-7cbb9b51177d.png)

![screenshot from 2017-12-04 16-53-39](https://user-images.githubusercontent.com/1210951/33563742-0560f0f0-d919-11e7-85f6-5db8d218c636.png)

![screenshot from 2017-12-04 16-57-15](https://user-images.githubusercontent.com/1210951/33563748-090a8d92-d919-11e7-8672-94136c7f8f4f.png)

**Comment by ttmc** [comment]
![screenshot from 2017-12-04 17-33-50](https://user-images.githubusercontent.com/1210951/33564207-3d2ca06e-d91a-11e7-8479-6f11202c8339.png)

**Comment by sbellem** [comment]
Could we disable this thing until it is fully worked out? PR https://github.com/bigchaindb/bigchaindb/pull/1899  has a pending check right now and clicking details leads to the same error as posted above:

```
Error

There is no CLA to sign for bigchaindb/bigchaindb

(Repository not found in Database)
```

**Comment by ttmc** [comment]
I disconnected CLA Assistant from the bigchaindb/bigchaindb repository on Monday before 6:00 pm, but apparently that didn't work either. It's even more broken than I thought.

**Comment by ttmc** [comment]
I looked around in the GitHub settings for the bigchaindb/bigchaindb repository, and the settings for the bigchaindb organization, to see if there were any remnants of CLA Assistant. I found one in the organization settings:

![screenshot from 2017-12-05 11-47-57](https://user-images.githubusercontent.com/1210951/33603362-910ae770-d9b2-11e7-8f04-6d4133235384.png)

I changed it to deny access.

**Comment by ttmc** [comment]
I also found CLA Assistant lurking in my personal GitHub settings, under "Applications - Authorized OAuth Apps". I revoked access.

![screenshot from 2017-12-05 11-57-54](https://user-images.githubusercontent.com/1210951/33603704-c2664908-d9b3-11e7-80d3-13305cc4885b.png)

**Comment by ttmc** [comment]
I created a new pull request this morning, on the bigchaindb/bigchaindb repo, and it doesn't have the CLA Assistant warning at the bottom, see https://github.com/bigchaindb/bigchaindb/pull/1933 Hopefully that will be true of _all_ new pull requests.

Unfortunately, all the bigchaindb/bigchaindb pull requests that existed at the time I did the CLA Assistant experiment were affected. I am sorry.
